pkg/webhook/resourceusage/util.go

```diff
@@ -0,0 +1,245 @@
+	"m":  1 << 20,
+	"gb": 1 << 30,
+	"g":  1 << 30,
+	"tb": 1 << 40,
+	"t":  1 << 40,
+	"pb": 1 << 50,
+	"p":  1 << 50,
+}
+
+var javaStringPattern = regexp.MustCompile("([0-9]+)([a-z]+)?")
+var javaFractionStringPattern = regexp.MustCompile("([0-9]+\\.[0-9]+)([a-z]+)?")
+
+// Logic copied from https://github.com/apache/spark/blob/5264164a67df498b73facae207eda12ee133be7d/common/network-common/src/main/java/org/apache/spark/network/util/JavaUtils.java#L276
+func parseJavaMemoryString(str string) (int64, error) {
+	lower := strings.ToLower(str)
+	if matches := javaStringPattern.FindStringSubmatch(lower); matches != nil {
+		value, err := strconv.ParseInt(matches[1], 10, 64)
+		if err != nil {
+			return 0, err
+		}
+		suffix := matches[2]
+		if multiplier, present := javaStringSuffixes[suffix]; present {
+			return multiplier * value, nil
+		}
+	} else if matches = javaFractionStringPattern.FindStringSubmatch(lower); matches != nil {
+		value, err := strconv.ParseFloat(matches[1], 64)
+		if err != nil {
+			return 0, err
+		}
+		suffix := matches[2]
+		if multiplier, present := javaStringSuffixes[suffix]; present {
+			return int64(float64(multiplier) * value), nil
+		}
+	}
+	return 0, fmt.Errorf("could not parse string '%s' as a Java-style memory value. Examples: 100kb, 1.5mb, 1g", str)
+}
+
+// Logic copied from https://github.com/apache/spark/blob/c4bbfd177b4e7cb46f47b39df9fd71d2d9a12c6d/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicDriverFeatureStep.scala
+func memoryRequiredForSparkPod(spec so.SparkPodSpec, memoryOverheadFactor *string, appType so.SparkApplicationType, replicas int64) (int64, error) {
+	var memoryBytes int64
+	if spec.Memory != nil {
+		memory, err := parseJavaMemoryString(*spec.Memory)
+		if err != nil {
+			return 0, err
+		}
+		memoryBytes = memory
+	} else {
+		memoryBytes = defaultMemoryBytes
+	}
+	var memoryOverheadBytes int64
+	if spec.MemoryOverhead != nil {
+		overhead, err := parseJavaMemoryString(*spec.MemoryOverhead)
+		if err != nil {
+			return 0, err
+		}
+		memoryOverheadBytes = overhead
+	} else {
+		var overheadFactor float64
+		if memoryOverheadFactor != nil {
+			overheadFactorScope, err := strconv.ParseFloat(*memoryOverheadFactor, 64)
+			if err != nil {
+				return 0, err
+			}
+			overheadFactor = overheadFactorScope
+		} else {
+			if appType == so.JavaApplicationType {
+				overheadFactor = defaultMemoryOverhead
+			} else {
+				overheadFactor = nonJvmDefaultMemoryOverhead
+			}
+		}
+		memoryOverheadBytes = int64(math.Max(overheadFactor*float64(memoryBytes), minMemoryOverhead))
+	}
+	return (memoryBytes + memoryOverheadBytes) * replicas, nil
+}
+
+func resourceUsage(spec so.SparkApplicationSpec) (ResourceList, error) {
+	driverMemoryOverheadFactor := spec.MemoryOverheadFactor
+	executorMemoryOverheadFactor := spec.MemoryOverheadFactor
+	if spec.Driver.MemoryOverhead != nil {
+		driverMemoryOverheadFactor = spec.Driver.SparkPodSpec.MemoryOverhead
```

Comment:
The unit of `MemoryOverheadFactor` (a fractional value, e.g., 0.1) is different than that of `MemoryOverhead` (MiBs).